External/FEXCore/Source/Interface/IR/Passes/DeadStoreElimination.cpp

```diff
@@ -0,0 +1,329 @@
+
+  return IsFPR(Offset);
+}
+
+
+
+uint64_t FPRBit(uint32_t Offset, uint32_t Size) {
+  if (!IsFPR(Offset)) {
+    return 0;
+  }
+
+  auto begin = offsetof(FEXCore::Core::ThreadState, State.xmm[0][0]);
+
+  auto regn = (Offset - begin)/16;
+  auto bitn = regn * 3;
+
+  if (!IsTrackedWriteFPR(Offset, Size))
+    return 7UL << (bitn);
+
+  if (Size == 16)
+    return  7UL << (bitn);
+  else if (Size == 8)
+    return  3UL << (bitn);
+  else if (Size == 4)
+    return  1UL << (bitn);
+  else
+    LogMan::Msg::A("Unexpected FPR size %d", Size);
+}
+
+struct Info {
+  FlagInfo flag;
+  GPRInfo gpr;
+  FPRInfo fpr;
+};
+
+
+/**
+ * @brief This is a temporary pass to detect simple multiblock dead flag/gpr/fpr stores
+ *
+ * First pass computes which flags/gprs/fprs are read and written per block
+ *
+ * Second pass computes which flags/gprs/fprs are stored, but overwritten by the next block(s).
+ * It also propagates this information a few times to catch dead flags/gprs/fprs across multiple blocks.
+ *
+ * Third pass removes the dead stores.
+ *
+ */
+bool DeadStoreElimination::Run(IREmitter *IREmit) {
+  std::unordered_map<OrderedNode*, Info> InfoMap;
+
+  bool Changed = false;
+  auto CurrentIR = IREmit->ViewIR();
+
+  // Pass 1
+  // Compute flags/gprs/fprs read/writes per block
+  // This is conservative and doesn't try to be smart about loads after writes
+  {
+    for (auto [BlockNode, BlockIROp] : CurrentIR.GetBlocks()) {
+      for (auto [CodeNode, IROp] : CurrentIR.GetCode(BlockNode)) {
+
+        //// Flags ////
+        if (IROp->Op == OP_STOREFLAG) {
+          auto Op = IROp->C<IR::IROp_StoreFlag>();
+          
+          auto& BlockInfo = InfoMap[BlockNode];
+
+          BlockInfo.flag.writes |= 1UL << Op->Flag;
+        } else if  (IROp->Op == OP_INVALIDATEFLAGS) {
+          auto Op = IROp->C<IR::IROp_InvalidateFlags>();
+
+          auto& BlockInfo = InfoMap[BlockNode];
+
+          BlockInfo.flag.writes |= Op->Flags;
+        } else if (IROp->Op == OP_LOADFLAG) {
+          auto Op = IROp->C<IR::IROp_LoadFlag>();
+
+          auto& BlockInfo = InfoMap[BlockNode];
+
+          BlockInfo.flag.reads |= 1UL << Op->Flag;
+        } else if (IROp->Op == OP_STORECONTEXT) {
+          auto Op = IROp->C<IR::IROp_StoreContext>();
+
+          auto& BlockInfo = InfoMap[BlockNode];
+
+          //// GPR ////
+          if (IsFullGPR(Op->Offset, IROp->Size))
+            BlockInfo.gpr.writes |= GPRBit(Op->Offset);
+          else
+            BlockInfo.gpr.reads |= GPRBit(Op->Offset);
+
+          //// FPR ////
+          if (IsTrackedWriteFPR(Op->Offset, IROp->Size))
+            BlockInfo.fpr.writes |= FPRBit(Op->Offset, IROp->Size);
+          else
+            BlockInfo.fpr.reads |= FPRBit(Op->Offset, IROp->Size);
+        } else if (IROp->Op == OP_STORECONTEXTINDEXED ||
+               IROp->Op == OP_LOADCONTEXTINDEXED) {
+          auto& BlockInfo = InfoMap[BlockNode];
+
+          //// GPR ////
+          // We can't track through these
```

Comment:
We probably could if we added a restriction that OP_STORECONTEXTINDEXED was only valid for x87, which is the only place we use it.

Reply:
It's also used for the segment registers afaik. @Sonicadvance1 thoughts?

Reply:
Hmm, we might.

Either way, as long as we can prove it's only used for for a limited set of registers we can avoid this optimisation killer.

Reply:
It's only used for 32bit and LDT/GDT loading of segment register data. Very minor.

Reply:
Created follow up https://github.com/FEX-Emu/FEX/issues/677